### DISCUSSION TITLE
Improve HookRegistry hook calling conventions

### DISCUSSION THREAD

**Discussion by asmecher** [pkp/pkp-lib#8089] [Accepted Proposals]
`HookRegistry` has some long-standing implementation quirks that make hook-based code hard to read and maintain:
- Hook parameters are be passed through an array and must be unpacked in sequence
- Type hinting is not possible within the parameter array
- Named parameters are not available within the parameter array
- The return value from a hook callback is perpetually confusing

Clean up these conventions (in a backwards-compatible way).

Proposal PR: https://github.com/pkp/pkp-lib/pull/8084

- The `HookRegistry` class has been renamed to `Hook`, with backwards-compatible class aliases provided.
- `Hook::add` is now preferred to `HookRegistry::register`. Backwards compatible.
- `Hook::run` is now preferred to `HookRegistry::call`. Backwards compatible -- but to emulate old calling conventions using a parameter array, wrap the parameters in `[]` a second time:
   ```php
   HookRegistry::run('someHookName', [[$mySubmission, $myPublication]]);
   ```

## Passing parameters to hook callbacks

`Hook::run` offers several new possibilities for callbacks:

### Callback function parameters properly declared

Previously, callbacks had to pass parameters via an array:

```php
HookRegistry::register('someHookName', function(string $hookName, array $args) {
    list($submission, $publication) = $args;
    ...
});

HookRegistry::call('someHookName', [$mySubmission, $myPublication]);
```

Using `Hook::run` instead, the parameters can be named directly, including type hinting:
```php
Hook::add('someHookName', function(string $hookName, Submission $submission, Publication $publication) {
    ...
});

Hook::run('someHookName', [$mySubmission, $myPublication]);
```

References are also supported; these can be declared in the callback parameter list as usual, and should be passed into the array as references as well (as before, even with objects):

```php
Hook::add('someHookName', function(string $hookName, Submission $submission, Publication &$publication) {
    ...
    $publication = $someOtherPublication;
});

Hook::run('someHookName', [$mySubmission, &$myPublication]);
```

## Named arguments

`Hook::run` supports named arguments; this can help keep code readable when longer lists of parameters are provided or parameter ordering might be unclear.

```php
Hook::add('someHookName', function(string $hookName, Submission $submission, Publication $publication) {
    ...
});

Hook::run('someHookName', ['submission' => $mySubmission, 'publication' => $myPublication]);
```

## Hook return conventions

In the past, hook callbacks returned `true` to interrupt later hooks, or `false` to continue. The meaning of these values was hard to remember. This change adds `Hook::ABORT` and `Hook::CONTINUE` constants to use instead. (This change is backwards compatible; the constants are equivalent to `true` and `false`.)

```php
Hook::add('someHookName', function(string $hookName) {
    return Hook::ABORT; // Stop processing subsequent hook registrants
});
```

**Comment by NateWr** [comment]
I like it! :+1: A few thoughts:

1. If we're deprecating and renaming things, what about shortening it slightly to `Hook` instead of `HookRegistry`. (`HookRegistry` can still be around for backwards compatibillity.)

3. Instead of `register` and `execute`, what about `Hook::add()` and `Hook::run()`?

4. Is aborting a hook necessary? I never could think of a good use case for this. I wonder if it's worth thinking about an approach where the result of a callback can be passed to the next callback. Something like:

```php
HookRegistry::register('my::hook', function(string $hookName, Submission $submission, $result = null) {
    return true;
});
HookRegistry::register('my::hook', function(string $hookName, Submission $submission, $result = null) {
    if ($result === true) {
        return $result;
    }
    return false;
});
```

Not loving that syntax, but you get the idea. A bit like `array_reduce()` or similar methods where the result accumulates over each iteration. (This is, for example, how middleware is implemented in some routers. Each middleware receives the request and response, and modifies the response, then the next middleware receives the modified response.) Maybe a special kind of hook could be used?

5. Maybe don't repeat "hook" in the constant names? So `HookRegistry::HOOK_SEQUENCE_CORE` could be `HookRegistry::SEQUENCE_CORE`.

6. The last big question for me is around documentation and discovery of hooks. I think that's been a consistent complaint over the years. How do people know what hooks are available and what parameters they receive? A couple of ideas:

We could use small static classes for hooks, which would give us code completion and type hinting in the editors. The downside is that we'd have a lot of small classes added to the codebase:

```php
use PKP\hook\Submission as HookSubmission;
HookRegistry::register(HookSubmission::add, function(Submission $submission) {});
HookRegistry::execute(HookSubmission::add($submission));
```

Or we could come up with a consistent docblock format. We would be able to generate documentation, but wouldn't get in-editor code completion:

```php
/**
* A submission has been added
*
* @param Submission $submission
*/
Hook::execute('Submission::add', [$submission]);
```

**Reply by asmecher** [reply]
Good ideas, @NateWr! I've adopted the following:
- Renaming `HookRegistry` class to `Hook`
- Renaming `Hook::execute` to `Hook::run`
- Renaming `Hook::register` to `Hook::add`
- Removed `HOOK_` prefix from constants

> Is aborting a hook necessary? I never could think of a good use case for this.

Yes, though it's not common (usually just letting OJS do its thing -- augmented by the plugin -- is the right behaviour).

Some examples where it's useful:
- The the lens galley plugin registers for `ArticleHandler::view::galley`, and on success the [callback returns true](https://github.com/asmecher/lensGalley/blob/52f79f18e1c2a90aab39e7fcdcee4889a4d614b6/LensGalleyPlugin.inc.php#L111) to [inhibit `ArticleHandler` from redirecting](https://github.com/pkp/ojs/blob/d2d3107fcefe3c3e709b3dd56fef7a7008587ee5/pages/article/ArticleHandler.inc.php#L356..L369).
- The citation style language plugin registers for `LoadHandler`, and when it detects that a request should be handled by the plugin, it [returns true](https://github.com/pkp/citationStyleLanguage/blob/8d4d3c7f57e5e23308921ff4495a0396abcc81ef/CitationStyleLanguagePlugin.inc.php#L660) to inhibit `PKPPageHandler` from [handling it itself, or generating a 404 on failure](https://github.com/pkp/pkp-lib/blob/main/classes/core/PKPPageRouter.inc.php#L247..L262).

I suspect the opacity of `return true` / `return false` have made the utility of this invisible! And in most plugins you can forget to return anything at all and get the desired default behaviour (hook registrants will continue to be processed).

As for code calling hooks, it should look like this when using the constants:

```php
if (Hook::run('MyHookName', [&$page, &$op, &$sourceFile, &$handler]) !== Hook::ABORT) {
   // here's the normal behaviour
}
```

...but of course it's up to the code that calls the hook to decide whether this is a "useful" pattern, i.e. whether there is some behaviour a plugin may want to replace. In many cases it'll just be a straight call to `Hook::run`, no conditional wrapped around it.

This is also useful when implemented with hook sequences. Consider for example a plugin that integrates a 3rd party statistics toolset by recording article page views. It would register for a "normal" hook sequence. The lens galley plugin registers for a ["late" hook sequence](https://github.com/asmecher/lensGalley/blob/52f79f18e1c2a90aab39e7fcdcee4889a4d614b6/LensGalleyPlugin.inc.php#L38). This combination ensures that the stats plugin would execute its callback *before* the lens galley stops hook processing and prevents OJS from continuing.

> I wonder if it's worth thinking about an approach where the result of a callback can be passed to the next callback...

This is available to registrants if they want to do it, but it requires all registrants to the hook to "play nice".

> The last big question for me is around documentation and discovery of hooks.

Agreed, this requires more thought and I don't have a good answer at the moment. I would favour a docblock approach to thousands of tiny classes. I suspect there is something out there that's solved this problem in another system -- I'll ask the tech committee if they have any ideas.

**Reply by nils-stefan-weiher** [reply]
> >   The last big question for me is around documentation and discovery of hooks.
> 
> Agreed, this requires more thought and I don't have a good answer at the moment. I would favour a docblock approach to thousands of tiny classes. I suspect there is something out there that's solved this problem in another system -- I'll ask the tech committee if they have any ideas.

Maybe this could be solved by a convention. The hooks are identified by strings, the first step would be to replace the string by a string constant. If every class, which has methods which calls hooks would use a public class constant like.

For example in the PublicationFormatForm class:
```
public const HOOK_EXECUTE = "PublicationFormatForm::execute";
```
At first sight this look like a lot of copy paste, but maybe this could be enhanced?

This could be used like this to register and run:
```
Hook::add(PublicationFormatForm::HOOK_EXECUTE, ...);

Hook::run(PublicationFormatForm::HOOK_EXECUTE, ... );
```
This would already help with discoverability and introduces the coupling to the hook calling class which is helpful for refactoring (if you want to find out which plugin hooks break when you change a class or method ).

The template hooks would have to be considered maybe purely by convention or by the corresponding handler class?
For example in class CatalogBookHandler:
```
public const HOOK_TEMPLATE_BOOK_FOOTER = "Templates::Catalog::Book::Footer::PageFooter"
```
Issues with embedded templates already spring to my mind while writing this.
Maybe template hooks have to be called by their path?
What do you think?

**Reply by NateWr** [reply]
I think it's a good idea. One thing we would need to be wary of is that someone could use `self` to call a hook:

```php
Hook::run(self::HOOK_EXECUTE, ...);
```

That might complicate efforts to discover all uses of a hook, for example. But I don't think it's a blocker.

**Reply by NateWr** [reply]
I wonder if we could use a method on the class itself, so that we get the code completion of methods without creating thousands of little classes. Something like:

```php
namespace APP\publication\Repository;

class Repository
{
    public function edit(Publication $publication, ...);
    {
        // ...
        Hook::run(self::HOOK_EDIT($publication));
    }
    public function HOOK_EDIT(Publication $publication)
    {
        return Hook::_call([Publication $publication]);
    }
}
```

```php
use APP\publication\Repository as PublicationRepo;

Hook::add(PublicationRepo::HOOK_EDIT, function(Publication $publication) {
    // ...
});
```

**Reply by asmecher** [reply]
We seem to have two leading options for hook documentation:

#### Hook names as class constants

This seems very promising, but I'd like to avoid having several representations for a hook name (e.g. `'MyClass::LITERAL_HOOK_NAME'` (string literal), `MyClass::HOOK_CONSTANT` (class constant), and within MyClass, `self::HOOK_CONSTANT`. It also introduces inheritance issues -- the constant can be accessed via subclasses (`MyClassDescendant::HOOK_CONSTANT`). I often have to find out where a hook is used (across many repos) and being able to grep for it easily is very helpful.

Here is a possibility...
- For legacy code, string literals are preferred. (Making them class constants with backwards-compatible values would invite use of both syntaxes, which would make it hard to grep for.)
- For new code, build the hook name in a class constant using magic constants:
    ```php
    class myClass {

        /**
         * Brief documentation goes here
         * No specific proposal for format, but it should probably include the callback parameter list
         */
        const HOOK_EXECUTE = self::class . '::execute';

        function execute() {
            Hook::call(self::HOOK_EXECUTE, function() {
                ...
            });
        }
    }
    ```
    - This makes it impossible to use the string formulation (unless you contort). It relies on class namespacing rather than `Some::Artificial::String::Heirarchy`.
    - There can be several ways of accessing the hook name constant if it is accessed via subclasses, and I don't think we can get around this except by convention -- hook name constants should always be accessed on the classes where they are defined.
    - There's still freedom to name hooks as desired within the class. Usually it'll be the function name, but we have `::before` and `::after` suffixes in many places, for example.
    - `self::class` always refers to the class as it's named where the constant is set -- it doesn't give a class alias or subclass name. (We have [toyed around with "magic" hook names](https://github.com/pkp/pkp-lib/blob/fa54a0ee0ade9989a6d5c4247d946e002b8da048/classes/form/Form.inc.php#L103) that try to be smarter than this -- but I think we should not use these. They are undocumentable, ungreppable, and confusing, and have lead to a lot of bad `parent::`ing conventions.)

#### Docblocks for hooks

I think docblocks are very promising. They would introduce better code hygiene for hooks at the point of relevance -- the head of the function where they're active. They introduce a place for concise documentation of the hook's purpose.

Here is an example of a tool to process `@hook` docblocks: https://github.com/matzeeable/wp-hookdoc

These could probably be mixed in with the above.

**Reply by NateWr** [reply]
One point of clarification: the proposal is not to use class constants but to use class methods. This allows them to be documented so that an IDE can pick up the function signature:

```php
Hook::add(ExampleForm::HOOK_EXECUTE, function(ExampleForm $form, Submission $submission) {
    // ...
});
```

In my IDE, I can read the full docblock from `ExampleForm::HOOK_EXECUTE` when I'm adding the hook.

That said, I do like docblocks and the wp hook documentation was my main inspiration for that approach. I'm not sure how easy it would be to repurpose wp-hookdoc for our needs but it certainly looks promising.

**Reply by asmecher** [reply]
Thanks, all! I think we have consensus on the [proposed `Hook` class changes](https://github.com/pkp/pkp-lib/pull/8084) discussed at the [head](https://github.com/pkp/pkp-lib/discussions/8089#discussion-4201246), but still need to work out some details regarding how we refer to hooks (by string, by constant, whether the constant would be built from `__CLASS__`, etc) and how hooks are documented (e.g. docblock and toolset). 

Are there any objections to merging the PR and working separately on the remaining parts? I would especially like to clean up code that refers to hook parameters by index; see e.g.: https://github.com/pkp/ops/pull/307#discussion_r936769832

**Reply by NateWr** [reply]
No objections here.

I went looking for ways to document the function arguments of a `callable` type and it doesn't look great. I found this discussion on [phpDocumentor](https://github.com/phpDocumentor/phpDocumentor/issues/1712) which suggests some tools are moving forward. Examples of [PHPStan](https://phpstan.org/writing-php-code/phpdoc-types#callables) and [Psalm](https://psalm.dev/docs/annotating_code/type_syntax/callable_types/).

These suggest that static analysis tools can pick up something like this:

```php
class Example
{
    /**
     * @param callable(int, string, array, Example):void
     */
    public static function HOOK_TEST(callable $cb)
    {
        return $cb(1, 'test', ['a', 'b']);
    }
}
```

However, it didn't work with VS Code. I'd expect when typing `Example::HOOK_TEST(function(` that I'd see the type hint for the callable, but no luck. I'd be curious to know if PHPStorm picks this up. @Vitaliy-1 maybe we can check on our call next week?

**Reply by asmecher** [reply]
:+1: I've committed the improvements from the draft PR above, and adaptations for the application repos and pkp-lib.

**Reply by asmecher** [reply]
Note that there are a couple of failed PHP RFCs addressing type-safe callables:
- https://wiki.php.net/rfc/typesafe-callable
- https://wiki.php.net/rfc/callable-types

**Reply by jonasraoni** [reply]
Given the [lack of priorities for the handlers](https://github.com/pkp/pkp-lib/discussions/8010#discussioncomment-3405247), I agreed to close the discussion to replace the hooks by the Laravel event system =]

About the type hinting/namespacing issue above, I would vote to use class based events (initial motivation here: https://github.com/pkp/pkp-lib/discussions/7979). Basically the Laravel event system with a priority queue.

**Reply by asmecher** [reply]
On the subject of hook documentation -- we discussed this again at the Hannover sprint.

We looked at two PHP self-documentation tools:
- Doxygen (which we currently use)
-  PHPDocumentor (which we are interested in migrating to)

Neither has clear tools to allow fishing through code for hook calls for automatic documentation.

PHPDocumentor has a (nascent) extension toolkit. It also has support for custom tags (https://github.com/phpDocumentor/ReflectionDocBlock/blob/master/examples/04-adding-your-own-tag.php).

Ideas:
- Support e.g. `@hook` in docblocks...
   ```
   /**
    * @hook My:Hook:Name Some hook description
    */
   ```
- Discourage the string-based assembly of hooks (these will always be inscrutable and hard to predict; naming them explicitly needs to be done consistently, but is concise and behaves predictably)
- Assess whether we want to move to PHPDocumentor
- Then look into an extension for Doxygen or PHPDocumentor that greps hook calls out of the code in a way that behaves equivalently to `@hook` tags -- ideally either would result in an entry in the self-doc.

**Reply by nils-stefan-weiher** [reply]
Looks good to me!

**Reply by asmecher** [reply]
There's a tool and a list of alternatives for WordPress hooks that would make a great starting point: https://github.com/pronamic/wp-documentor#alternatives

**Reply by asmecher** [reply]
Step 1: Get the php-cs-fixer to add `@hook` self-documentation to function docblocks:
- pkp-lib: https://github.com/pkp/pkp-lib/pull/9338
- ojs: https://github.com/pkp/ojs/pull/4049
- omp: https://github.com/pkp/omp/pull/1475
- ops: https://github.com/pkp/ops/pull/582
(The re-linting commit is kept separate from the tooling change.)

To generate `@hook` self-documentation in PHP classes, use...
- Application: `lib/pkp/lib/vendor/bin/php-cs-fixer fix --allow-risky=yes --path-mode=intersection --config=.php-cs-fixer.php --using-cache=no -vvv .`
- pkp-lib: `./lib/vendor/bin/php-cs-fixer fix --allow-risky=yes --path-mode=intersection --config=.php-cs-fixer.php  --using-cache=no -vvv .`

Note that these *only* add `@hook` self-documentation where it's missing! It won't correct/re-generate hook self-docs when they already exist.

Annotations look like this:
- `@hook API::stats::issues::params [[&$allowedParams, $slimRequest]]`
  - The `[[` ... `]]` indicates a nested array, which is the old Hook::call convention.
  - The `&` indicates that there's a reference on `$allowedParams`.
  - The corresponding function to accept the hook might look like:
     ```php
     function issueStatsParamsHookHandler(string $hookName, array $args): bool
     {
       $allowedParams =& $args[0]; // this was passed by reference and can be modified
       $slimRequest = $args[1];
       ...
       return Hook::CONTINUE; // or Hook::ABORT;
     }
     ```
- `@hook Some::Other::Hook ['param1' => &$param1, 'param2' => $param2]`
  - The `[` ... `]` is not a nested array, indicating the new Hook::run convention.
  - Named parameters are used. (We don't have to use these, but they're nice for long parameter lists. Otherwise positional parameters will be used instead.)
  - The corresponding function to accept the hook might look like:
     ```php
     function someOtherHookHandler(string $hookName, string &$param1, string $param2): bool
     {
       ...
       return Hook::CONTINUE; // or Hook::ABORT;
     }
     ```

**Reply by asmecher** [reply]
Step 2: Generate a human-readable hook list.

It turns out there was already a hook listing tool -- `lib/pkp/tools/getHooks.php` -- that dumped a JSON-encoded bare list of hooks by name.

I've extended that in [the PRs above](https://github.com/pkp/pkp-lib/discussions/8089#discussioncomment-7139458) to generate more information, and in multiple formats, including RST (which is supported -- barely -- by PHP's [hand-written documentation](https://docs.phpdoc.org/3.0/guide/hand-written-docs/index.html#hand-written-documentation) toolkit).

Result: A list of hooks added to the phpDocumentor output, with a link to it from the sidebar:
![image](https://github.com/pkp/pkp-lib/assets/200411/84af0cc8-fd5e-4e49-8abe-c457e57ecc2e)

This could use improvement, such as links from the filenames into that part of the reference, but the hand-written documentation feature seems very much to be a work in progress.

The normal phpDocumentor output also lists custom tags in context with function self-documentation:
![image](https://github.com/pkp/pkp-lib/assets/200411/cc66bf00-2241-40be-8677-b5239d6f5e02)

It's got some room to improve, but I think this is ready to launch.

**Reply by asmecher** [reply]
All merged! We have a hook list that can be generated in each application by...
```
# Update the hook list in PHPDocumentor's guide
php lib/pkp/tools/getHooks.php -r > docs/dev/guide/hooks.rst

# Re-generate PHPDocumentation
php phpDocumentor.phar

# Documentation will now be available in docs/dev/build
```

**Comment by asmecher** [comment]
Just capturing some thoughts in case I decide to push them further. Now that we have [first class callable syntax](https://www.php.net/manual/en/functions.first_class_callable_syntax.php) in PHP, we can already use [callables for hook registrants](https://github.com/pkp/pkp-lib/commit/e5f5f14ef7f180c485d90b55a718e61f5e0364f5). However, we may also consider using first-class callables for hook naming.

Rather than registering and calling hooks as strings...
```php
// Adding a hook registrant
Hook::add('MyClass::myHookName', fn() => {...});

// Running a hook, usually done within the MyClass::myHookName function
Hook::run('MyClass::myHookName');
```
...we may want to use callables interchangeably with strings:
```php
// Adding a hook registrant
Hook::add(MyClass::myHookName(...), fn() => {...});

// Running a hook, usually done within the MyClass::myHookName function
Hook::run($this->myHookName(...));
// ...or maybe...
Hook::run(__METHOD__); // not actually a first class callable, but also avoids repeating the method name
```
Experimentally, this can be accomplished with the following:
```diff
diff --git a/classes/plugins/Hook.php b/classes/plugins/Hook.php
index e425759b1b..4103875af2 100644
--- a/classes/plugins/Hook.php
+++ b/classes/plugins/Hook.php
@@ -31,18 +31,19 @@ class Hook
     /**
      * Get the current set of hook registrations.
      *
-     * @param string $hookName Name of hook to optionally return
+     * @param string $hook Hook to optionally return
      *
-     * @return mixed Array of all hooks or just those attached to $hookName, or
-     *   null if nothing has been attached to $hookName
+     * @return mixed Array of all hooks or just those attached to $hook, or
+     *   null if nothing has been attached to $hook
      */
-    public static function &getHooks(?string $hookName = null): ?array
+    public static function &getHooks(callable|string|null $hook = null): ?array
     {
         $hooks = & Registry::get('hooks', true, []);
-        if ($hookName === null) {
+        if ($hook === null) {
             return $hooks;
         }
 
+        $hookName = self::possibleCallableToString($hook);
         if (isset($hooks[$hookName])) {
             return $hooks[$hookName];
         }
@@ -52,23 +53,25 @@ class Hook
     }
 
     /**
-     * Clear hooks registered against the given name.
+     * Clear registrants from the specified hook.
      */
-    public static function clear(string $hookName): void
+    public static function clear(callable|string $hook): void
     {
+        $hookName = self::possibleCallableToString($hook);
         $hooks = & static::getHooks();
         unset($hooks[$hookName]);
     }
 
     /**
-     * Register a hook against the given hook name.
+     * Add a registrant against the given hook.
      *
-     * @param string $hookName Name of hook to register against
+     * @param string $hook Hook to register against
      * @param callable $callback Callback pseudo-type
      * @param int $hookSequence Optional hook sequence specifier SEQUENCE_...
      */
-    public static function add(string $hookName, callable $callback, int $hookSequence = self::SEQUENCE_NORMAL): void
+    public static function add(callable|string $hook, callable $callback, int $hookSequence = self::SEQUENCE_NORMAL): void
     {
+        $hookName = self::possibleCallableToString($hook);
         $hooks = & static::getHooks();
         $hooks[$hookName][$hookSequence][] = $callback;
     }
@@ -114,20 +117,21 @@ class Hook
     }
 
     /**
-     * Call each callback registered against $hookName in sequence.
+     * Call each callback registered against $hook in sequence.
      * The first callback that returns ABORT will interrupt processing and
      * this function will return ABORT; otherwise, all callbacks will be
      * called in sequence and the return value of this call will be
      * CONTINUE.
      *
      * The signature of a callback function should be:
-     *   function callback($hookName, ...) : bool;
+     *   function callback($hook, ...) : bool;
      * where ... corresponds to the parameters named/listed in the $args
      * parameter to Hook::call. These may be named if desired,
      * and may include references.
      */
-    public static function run(string $hookName, $args): bool
+    public static function run(callable|string $hook, $args): bool
     {
+        $hookName = self::possibleCallableToString($hook);
         $hooks = & static::getHooks();
         if (!isset($hooks[$hookName])) {
             return self::CONTINUE;
@@ -137,7 +141,7 @@ class Hook
         foreach ($hooks[$hookName] as $priority => $hookList) {
             foreach ($hookList as $callback) {
                 $params = array_merge([$hookName], $args);
-                if (call_user_func_array($callback, array_merge([$hookName], $args)) === self::ABORT) {
+                if (call_user_func_array($callback, array_merge([$hook], $args)) === self::ABORT) {
                     return self::ABORT;
                 }
             }
@@ -195,6 +199,14 @@ class Hook
         static $calledHooks = [];
         return $calledHooks;
     }
+
+    protected static function possibleCallableToString(callable|string|null $callableOrString): string
+    {
+        if (is_string($callableOrString)) return $callableOrString;
+
+        $rf = new \ReflectionFunction($callableOrString);
+        return get_class($rf->getClosureThis()) . '::' . $rf->getName();
+    }
 }
 
 if (!PKP_STRICT_MODE) {
```
This might move us ahead in our quest for proper hook self-documentation and static analysis, and reduces the frequent repetition of class and function names when calling hooks that are often named identically to the current function.
